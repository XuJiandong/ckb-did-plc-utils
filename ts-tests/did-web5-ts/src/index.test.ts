import * as cbor from "@ipld/dag-cbor";
import {
  hexFrom,
  Transaction,
  Hex,
  hashTypeId,
  bytesFrom,
  WitnessArgs,
  numFrom,
  OutPoint,
  CellOutput,
  Cell,
} from "@ckb-ccc/core";
import { readFileSync } from "fs";
import {
  DEFAULT_SCRIPT_ALWAYS_SUCCESS,
  Resource,
  Verifier,
} from "ckb-testtool";
import path from "path";
import { molecule, plc } from "./index";
import * as uint8arrays from "uint8arrays";
import { runCoverage } from "./coverage";
import { P256Keypair, Secp256k1Keypair } from "@atproto/crypto";

if (process.env.CKB_COVERAGE) {
  console.log(
    "The environment variable CKB_COVERAGE is defined. It's for coverage test only",
  );
}

export const DEFAULT_SCRIPT = path.join(
  __dirname,
  process.env.CKB_COVERAGE
    ? "../../../build/debug/did-web5-ts"
    : "../../../build/release/did-web5-ts",
);

export const DEFAULT_SCRIPT_HEX = hexFrom(readFileSync(DEFAULT_SCRIPT));
export const ALWAYS_SUCCESS_HEX = hexFrom(
  readFileSync(DEFAULT_SCRIPT_ALWAYS_SUCCESS),
);

function newLocalId(binaryDid: Hex): Hex {
  let did = bytesFrom(binaryDid);
  let str = "did:plc:" + uint8arrays.toString(did, "base32");
  return hexFrom(uint8arrays.fromString(str, "utf8"));
}

function jsonify(obj: any): any {
  return JSON.parse(
    JSON.stringify(
      obj,
      (_, value) =>
        typeof value === "bigint" ? `0x${value.toString(16)}` : value,
      2,
    ),
  );
}

async function main(
  result: plc.PlcOperationResult,
  config: {
    inputCellCount?: number;
    outputCellCount?: number;
    noAssociatePlc?: boolean;
    update?: boolean;
    updateLocalId?: boolean;
    invalidSignature?: boolean;
    shortArgs?: boolean;
    invalidCbor?: boolean;
    mismatchedHistory?: boolean;
  },
  shouldFail?: boolean,
): Promise<number> {
  const resource = Resource.default();
  const tx = Transaction.default();
  const script = resource.deployCell(DEFAULT_SCRIPT_HEX, tx, false);
  const alwaysSuccessScript = resource.deployCell(
    ALWAYS_SUCCESS_HEX,
    tx,
    false,
  );
  let transferredFrom: Hex | null = null;
  let codeHashToRun: Hex | null = null;

  // cell data
  if (config?.noAssociatePlc) {
    transferredFrom = null;
  } else {
    transferredFrom = newLocalId(result.binaryDid);
  }
  // When testing invalid CBOR scenarios, use "0x82" which represents a CBOR array
  // expecting 2 elements but provides none, making it invalid CBOR format
  let cborData = config?.invalidCbor ? bytesFrom("0x82") : cbor.encode("");
  let didWeb5Data = molecule.DidWeb5Data.from({
    value: {
      document: cborData,
      localId: transferredFrom,
    },
  });

  if (config?.update || config?.updateLocalId) {
    // script args
    let typeScript = script.clone();
    typeScript.args = hexFrom("0x" + "0".repeat(40));
    codeHashToRun = typeScript.hash();
    const inputCell = resource.mockCell(
      alwaysSuccessScript,
      typeScript,
      hexFrom(didWeb5Data.toBytes()),
    );
    // input cells
    for (let i = 0; i < (config?.inputCellCount ?? 1); i++) {
      tx.inputs.push(Resource.createCellInput(inputCell));
    }

    // output cells
    for (let i = 0; i < (config?.outputCellCount ?? 1); i++) {
      tx.outputs.push(
        Resource.createCellOutput(alwaysSuccessScript, typeScript),
      );
      let newDidWeb5Data = didWeb5Data.clone();
      if (config?.updateLocalId) {
        newDidWeb5Data.value.localId = hexFrom(newLocalId("0x00"));
      } else {
        newDidWeb5Data.value.document = hexFrom(
          cbor.encode({ key: "hello, world" }),
        );
      }
      tx.outputsData.push(hexFrom(newDidWeb5Data.toBytes()));
    }
  } else {
    // input cells
    const inputCell = resource.mockCell(alwaysSuccessScript);
    tx.inputs.push(Resource.createCellInput(inputCell));

    // Note: Type script args must be computed after input cells are added to the transaction
    // because the type ID depends on the first input cell's outpoint
    let typeScript = script.clone();
    let typeId = hashTypeId(tx.inputs[0], 0);
    typeScript.args = hexFrom(typeId.slice(0, config?.shortArgs ? 10 : 42)); // 20 bytes Type ID
    codeHashToRun = typeScript.hash();

    let count = config?.outputCellCount ?? 1;
    for (let i = 0; i < count; i++) {
      tx.outputs.push(
        Resource.createCellOutput(alwaysSuccessScript, typeScript),
      );
      tx.outputsData.push(hexFrom(didWeb5Data.toBytes()));
    }
  }

  // witness
  if (!config?.noAssociatePlc) {
    let txHash = tx.hash();
    if (config.invalidSignature) {
      result.rotationKeyIndices.push(0n);
      result.sig = "0x00";
    } else {
      await plc.signDidWeb5(result, 0, txHash);
    }
    if (!result.sig) {
      throw new Error("Signature is required");
    }
    let web5Witness = molecule.DidWeb5Witness.from({
      localIdAuthorization: {
        history: result.history,
        sig: result.sig,
        rotationKeyIndices: result.rotationKeyIndices,
      },
    });
    let witnessArgs = WitnessArgs.from({
      outputType: web5Witness.toBytes(),
    });
    tx.setWitnessArgsAt(0, witnessArgs);
  }

  const verifier = Verifier.from(resource, tx);
  if (process.env.CKB_COVERAGE) {
    const txFile = JSON.stringify(verifier.txFile());
    let count = config?.outputCellCount ?? 1;
    runCoverage(
      "type",
      count === 0 ? "input" : "output",
      0,
      txFile,
      shouldFail ?? false,
    );
    return 0;
  } else {
    if (shouldFail) {
      await verifier.verifyFailure(undefined, false, {
        codeHash: codeHashToRun,
      });
      return 0;
    } else {
      return verifier.verifySuccess(true, { codeHash: codeHashToRun });
    }
  }
}

describe("did-web5-ts", () => {
  test("it should process a genesis operation without associated did:plc correctly", async () => {
    let result = await plc.generateOperations();
    await main(result, { noAssociatePlc: true });
  });
  test("it should process a genesis operation with associated did:plc correctly", async () => {
    let result = await plc.generateOperations();
    await main(result, {});
  });
  test("it should reject invalid cbor format", async () => {
    let result = await plc.generateOperations();
    await main(result, { invalidCbor: true }, true);
  });

  test("it should reject process invalid args (!= 20 bytes)", async () => {
    let result = await plc.generateOperations();
    await main(result, { shortArgs: true }, true);
  });
  test("it should reject multiple output cells while minting", async () => {
    let result = await plc.generateOperations();
    await main(result, { outputCellCount: 2 }, true);
  });
  test("it should process several operations with associated did:plc correctly", async () => {
    let result = await plc.generateOperations({ moreOps: true });
    await main(result, {});
  });
  test("it should process an update correctly", async () => {
    let result = await plc.generateOperations();
    await main(result, { update: true });
  });
  test("it should reject an update with multiple outputs", async () => {
    let result = await plc.generateOperations();
    await main(result, { update: true, outputCellCount: 2 }, true);
  });
  test("it should reject an update with multiple inputs", async () => {
    let result = await plc.generateOperations();
    await main(result, { update: true, inputCellCount: 2 }, true);
  });
  test("it should process an update to burn", async () => {
    let result = await plc.generateOperations();
    await main(result, { update: true, outputCellCount: 0 });
  });
  test("it should reject an update with local id changed", async () => {
    let result = await plc.generateOperations();
    await main(result, { updateLocalId: true }, true);
  });
  test("it should reject a genesis operation with associated did:plc and invalid signature", async () => {
    let result = await plc.generateOperations();
    await main(result, { invalidSignature: true }, true);
  });
  test("it should reject a genesis operation with associated did:plc and invalid signature 2", async () => {
    let result = await plc.generateOperations({ invalidSignature: true });
    await main(result, {}, true);
  });
  test("it should reject an operation with mismatched history length", async () => {
    let result = await plc.generateOperations({ mismatchedHistory: true });
    await main(result, {}, true);
  });

  test("it should re-create the spec example", async () => {
    let previousTxHash: Hex =
      "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559";
    const did = "0x8434cfe81aa825c275d513eee20e4235294e3420";

    const resource = Resource.default();
    resource.cellOutpointHash = previousTxHash;
    const baseTx = Transaction.default();
    const typeScript = resource.deployCell(DEFAULT_SCRIPT_HEX, baseTx, true);
    typeScript.args = did;
    const alwaysSuccessScript = resource.deployCell(
      ALWAYS_SUCCESS_HEX,
      baseTx,
      true,
    );

    const doc0 = {
      verificationMethods: {
        atproto: "did:key:zSigningKey",
      },
      alsoKnownAs: ["at://alice.test"],
      services: {
        atproto_pds: {
          type: "AtprotoPersonalDataServer",
          endpoint: "https://example.test",
        },
      },
    };
    const doc1 = {
      verificationMethods: {
        atproto: "did:key:zSigningKey",
      },
      alsoKnownAs: ["at://bob.test"],
      services: {
        atproto_pds: {
          type: "AtprotoPersonalDataServer",
          endpoint: "https://example.test",
        },
      },
    };

    const didWeb5Data0 = molecule.DidWeb5Data.from({
      value: {
        document: cbor.encode(doc0),
        localId: null,
      },
    });
    const didWeb5Data1 = molecule.DidWeb5Data.from({
      value: {
        document: cbor.encode(doc1),
        localId: null,
      },
    });

    expect(didWeb5Data0).toMatchSnapshot("data0");
    expect(didWeb5Data1).toMatchSnapshot("data1");

    // Creation Example
    {
      const tx = Transaction.default();
      tx.cellDeps = baseTx.cellDeps;
      const inputCell = resource.mockCell(alwaysSuccessScript);
      tx.inputs.push(Resource.createCellInput(inputCell));

      const typeId = hashTypeId(tx.inputs[0], 0);
      typeScript.args = hexFrom(typeId.slice(0, 42)); // 20 bytes Type ID
      expect(hexFrom(typeId.slice(0, 42))).toBe(did);

      tx.outputs.push(
        Resource.createCellOutput(
          alwaysSuccessScript,
          typeScript,
          numFrom(600),
        ),
      );
      tx.outputsData.push(hexFrom(didWeb5Data0.toBytes()));

      const verifier = Verifier.from(resource, tx);
      const txJson = jsonify(tx);

      expect(tx.hash()).toMatchSnapshot("creation-tx-hash");
      expect(txJson).toMatchSnapshot("creation-tx");
      verifier.verifySuccess(true, { codeHash: typeScript.hash() });

      previousTxHash = tx.hash();
    }

    // Update Example
    {
      const tx = Transaction.default();
      tx.cellDeps = baseTx.cellDeps;

      const inputCell = new Cell(
        new OutPoint(previousTxHash, numFrom(0)),
        new CellOutput(numFrom(600), alwaysSuccessScript, typeScript),
        hexFrom(didWeb5Data0.toBytes()),
      );
      resource.cells.set(inputCell.outPoint.toBytes().toString(), inputCell);
      tx.inputs.push(Resource.createCellInput(inputCell));

      tx.outputs.push(
        Resource.createCellOutput(
          alwaysSuccessScript,
          typeScript,
          numFrom(599),
        ),
      );
      tx.outputsData.push(hexFrom(didWeb5Data1.toBytes()));

      const verifier = Verifier.from(resource, tx);
      const txJson = jsonify(tx);

      expect(tx.hash()).toMatchSnapshot("update-tx-hash");
      expect(txJson).toMatchSnapshot("update-tx");
      verifier.verifySuccess(true, { codeHash: typeScript.hash() });

      previousTxHash = tx.hash();
    }

    // Deactivation Example
    {
      const tx = Transaction.default();
      tx.cellDeps = baseTx.cellDeps;

      const inputCell = new Cell(
        new OutPoint(previousTxHash, numFrom(0)),
        new CellOutput(numFrom(599), alwaysSuccessScript, typeScript),
        hexFrom(didWeb5Data1.toBytes()),
      );
      resource.cells.set(inputCell.outPoint.toBytes().toString(), inputCell);
      tx.inputs.push(Resource.createCellInput(inputCell));

      tx.outputs.push(
        Resource.createCellOutput(alwaysSuccessScript, undefined, numFrom(598)),
      );
      tx.outputsData.push("0x");

      const verifier = Verifier.from(resource, tx);
      const txJson = jsonify(tx);

      expect(txJson).toMatchSnapshot("deactivation-tx");
      verifier.verifySuccess(true, { codeHash: typeScript.hash() });

      previousTxHash = tx.hash();
    }
  });

  test("it should re-create the local id extension example", async () => {
    const previousTxHash: Hex =
      "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559";
    const did = "0x8434cfe81aa825c275d513eee20e4235294e3420";

    const key = await Secp256k1Keypair.import(
      "32288e73ba3cc99b0d2499f20c08cf4ef90d4f4f914ce21221d81a386bbc7e44",
    );
    const rotationKey1 = await Secp256k1Keypair.import(
      "1f8b611642074eac6733331e2120de2a83e983775391706649abefc2995bae73",
    );
    let rotationKey2 = await P256Keypair.import(
      "e7d9916064bcf1c5e80ef5882f84171dc0a46fb7c40f5b29a671f9155e172482",
    );

    const migration = await plc.generateOperations({
      key,
      rotationKey1,
      rotationKey2,
    });
    const localId = newLocalId(migration.binaryDid);
    expect(uint8arrays.toString(bytesFrom(localId))).toMatchSnapshot(
      "local-id",
    );

    const resource = Resource.default();
    resource.cellOutpointHash = previousTxHash;
    const tx = Transaction.default();
    const typeScript = resource.deployCell(DEFAULT_SCRIPT_HEX, tx, true);
    typeScript.args = did;
    const alwaysSuccessScript = resource.deployCell(
      ALWAYS_SUCCESS_HEX,
      tx,
      true,
    );

    const doc = {
      verificationMethods: {
        atproto: "did:key:zSigningKey",
      },
      alsoKnownAs: ["at://alice.test"],
      services: {
        atproto_pds: {
          type: "AtprotoPersonalDataServer",
          endpoint: "https://example.test",
        },
      },
    };
    const didWeb5Data = molecule.DidWeb5Data.from({
      value: {
        document: cbor.encode(doc),
        localId,
      },
    });

    // Migration Example
    const inputCell = resource.mockCell(alwaysSuccessScript);
    tx.inputs.push(Resource.createCellInput(inputCell));

    const typeId = hashTypeId(tx.inputs[0], 0);
    typeScript.args = hexFrom(typeId.slice(0, 42)); // 20 bytes Type ID
    expect(hexFrom(typeId.slice(0, 42))).toBe(did);

    tx.outputs.push(
      Resource.createCellOutput(alwaysSuccessScript, typeScript, numFrom(600)),
    );
    tx.outputsData.push(hexFrom(didWeb5Data.toBytes()));

    await plc.signDidWeb5(migration, 0, tx.hash());
    expect(migration.sig).toBeTruthy();
    if (!migration.sig) {
      throw new Error("Signature is required");
    }
    expect(cbor.decode(bytesFrom(migration.history[0]!))).toMatchSnapshot(
      "plc-genesis-op",
    );
    const web5Witness = molecule.DidWeb5Witness.from({
      localIdAuthorization: {
        history: migration.history,
        sig: migration.sig,
        rotationKeyIndices: migration.rotationKeyIndices,
      },
    });
    expect(jsonify(web5Witness)).toMatchSnapshot("witness");
    let witnessArgs = WitnessArgs.from({
      outputType: web5Witness.toBytes(),
    });
    tx.setWitnessArgsAt(0, witnessArgs);
    const verifier = Verifier.from(resource, tx);
    const txJson = jsonify(tx);

    expect(tx.hash()).toMatchSnapshot("local-id-migration-tx-hash");
    expect(txJson).toMatchSnapshot("local-id-migration-tx");
    verifier.verifySuccess(true, { codeHash: typeScript.hash() });
  });
});
