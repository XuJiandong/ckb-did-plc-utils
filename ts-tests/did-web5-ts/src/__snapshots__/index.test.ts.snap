// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`did-web5-ts it should re-create the local id extension example: local-id 1`] = `"web5:plc:bxvfvvygwbcnbmknn73t6pbu"`;

exports[`did-web5-ts it should re-create the local id extension example: local-id-migration-tx 1`] = `
{
  "cellDeps": [
    {
      "depType": "code",
      "outPoint": {
        "index": "0x0",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
    {
      "depType": "code",
      "outPoint": {
        "index": "0x1",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
  ],
  "headerDeps": [],
  "inputs": [
    {
      "previousOutput": {
        "index": "0x2",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
      "since": "0x0",
    },
  ],
  "outputs": [
    {
      "capacity": "0x258",
      "lock": {
        "args": "0x",
        "codeHash": "0x3dc9cdb5b2dcff8d4999004406558d4456ecf0f14c757ca353684c0bd4d3b50d",
        "hashType": "type",
      },
      "type": {
        "args": "0x8434cfe81aa825c275d513eee20e4235294e3420",
        "codeHash": "0xb95123c71a870e3f0f74a7ee1dab8268dbfbc1407b46733ebd1b41f854b4324a",
        "hashType": "type",
      },
    },
  ],
  "outputsData": [
    "0x00000000d90000000c000000b4000000a4000000a3687365727669636573a16b617470726f746f5f706473a264747970657819417470726f746f506572736f6e616c4461746153657276657268656e64706f696e747468747470733a2f2f6578616d706c652e746573746b616c736f4b6e6f776e4173816f61743a2f2f616c6963652e7465737473766572696669636174696f6e4d6574686f6473a167617470726f746f736469643a6b65793a7a5369676e696e674b657921000000776562353a706c633a62787666767679677762636e626d6b6e6e37337436706275",
  ],
  "version": "0x0",
  "witnesses": [
    "0x4c0200001000000010000000100000003802000038020000080000003002000010000000e60100002a020000d601000008000000ca010000a76373696778564c2d6b72665f6f385f4b4a5754362d51316532572d52466252443777313564575f4e4b726b6d5f634c2d355147346b5f70736270714a5f314f48626a3034716a70764f70724143507357384765582d6e5253433030416470726576f664747970656d706c635f6f7065726174696f6e687365727669636573a16b617470726f746f5f706473a264747970657819417470726f746f506572736f6e616c4461746153657276657268656e64706f696e747368747470733a2f2f6578616d706c652e636f6d6b616c736f4b6e6f776e4173817661743a2f2f616c6963652e6578616d706c652e636f6d6c726f746174696f6e4b6579738278396469643a6b65793a7a5133736859314d5351545a6f584431505370616a566237487642734841754a69455039354271426b5358316d6365794278396469643a6b65793a7a446e616570684464367a5a4c433353397873536e6b7579784b39316d62663467654e53666e6a334c715a65666158693373766572696669636174696f6e4d6574686f6473a167617470726f746f78396469643a6b65793a7a513373686e6d517963716d38344d684c48544a6675614a6843326a66774a69597a697a784c6162574e466269796d515440000000a3d767cb71da12c25e4a32366990600f64b0683934034982961cfc439d172d87797523c5d93207889bdfd3045bdff75d96d7294307011afb9f198f152f0347e2020000000000",
  ],
}
`;

exports[`did-web5-ts it should re-create the local id extension example: local-id-migration-tx-hash 1`] = `"0xf7d3002ebd20f694fa974a0d680607471c48a526887fab62bba2908af3c507b2"`;

exports[`did-web5-ts it should re-create the local id extension example: plc-genesis-op 1`] = `
{
  "alsoKnownAs": [
    "at://alice.example.com",
  ],
  "prev": null,
  "rotationKeys": [
    "did:key:zQ3shY1MSQTZoXD1PSpajVb7HvBsHAuJiEP95BqBkSX1mceyB",
    "did:key:zDnaephDd6zZLC3S9xsSnkuyxK91mbf4geNSfnj3LqZefaXi3",
  ],
  "services": {
    "atproto_pds": {
      "endpoint": "https://example.com",
      "type": "AtprotoPersonalDataServer",
    },
  },
  "sig": "L-krf_o8_KJWT6-Q1e2W-RFbRD7w15dW_NKrkm_cL-5QG4k_psbpqJ_1OHbj04qjpvOprACPsW8GeX-nRSC00A",
  "type": "plc_operation",
  "verificationMethods": {
    "atproto": "did:key:zQ3shnmQycqm84MhLHTJfuaJhC2jfwJiYzizxLabWNFbiymQT",
  },
}
`;

exports[`did-web5-ts it should re-create the local id extension example: witness 1`] = `
{
  "localIdAuthorization": {
    "history": [
      "0xa76373696778564c2d6b72665f6f385f4b4a5754362d51316532572d52466252443777313564575f4e4b726b6d5f634c2d355147346b5f70736270714a5f314f48626a3034716a70764f70724143507357384765582d6e5253433030416470726576f664747970656d706c635f6f7065726174696f6e687365727669636573a16b617470726f746f5f706473a264747970657819417470726f746f506572736f6e616c4461746153657276657268656e64706f696e747368747470733a2f2f6578616d706c652e636f6d6b616c736f4b6e6f776e4173817661743a2f2f616c6963652e6578616d706c652e636f6d6c726f746174696f6e4b6579738278396469643a6b65793a7a5133736859314d5351545a6f584431505370616a566237487642734841754a69455039354271426b5358316d6365794278396469643a6b65793a7a446e616570684464367a5a4c433353397873536e6b7579784b39316d62663467654e53666e6a334c715a65666158693373766572696669636174696f6e4d6574686f6473a167617470726f746f78396469643a6b65793a7a513373686e6d517963716d38344d684c48544a6675614a6843326a66774a69597a697a784c6162574e466269796d5154",
    ],
    "sig": "0xa3d767cb71da12c25e4a32366990600f64b0683934034982961cfc439d172d87797523c5d93207889bdfd3045bdff75d96d7294307011afb9f198f152f0347e2",
    "signingKeys": [
      "0x0",
      "0x0",
    ],
  },
}
`;

exports[`did-web5-ts it should re-create the spec example: creation-tx 1`] = `
{
  "cellDeps": [
    {
      "depType": "code",
      "outPoint": {
        "index": "0x0",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
    {
      "depType": "code",
      "outPoint": {
        "index": "0x1",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
  ],
  "headerDeps": [],
  "inputs": [
    {
      "previousOutput": {
        "index": "0x2",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
      "since": "0x0",
    },
  ],
  "outputs": [
    {
      "capacity": "0x258",
      "lock": {
        "args": "0x",
        "codeHash": "0x3dc9cdb5b2dcff8d4999004406558d4456ecf0f14c757ca353684c0bd4d3b50d",
        "hashType": "type",
      },
      "type": {
        "args": "0x8434cfe81aa825c275d513eee20e4235294e3420",
        "codeHash": "0xb95123c71a870e3f0f74a7ee1dab8268dbfbc1407b46733ebd1b41f854b4324a",
        "hashType": "type",
      },
    },
  ],
  "outputsData": [
    "0x00000000b40000000c000000b4000000a4000000a3687365727669636573a16b617470726f746f5f706473a264747970657819417470726f746f506572736f6e616c4461746153657276657268656e64706f696e747468747470733a2f2f6578616d706c652e746573746b616c736f4b6e6f776e4173816f61743a2f2f616c6963652e7465737473766572696669636174696f6e4d6574686f6473a167617470726f746f736469643a6b65793a7a5369676e696e674b6579",
  ],
  "version": "0x0",
  "witnesses": [],
}
`;

exports[`did-web5-ts it should re-create the spec example: creation-tx-hash 1`] = `"0xe54bb705cea6e500767ed4c0ba9ee92cad71e6b39c55283591147c5d10cc12b5"`;

exports[`did-web5-ts it should re-create the spec example: data0 1`] = `
Extended {
  "type": "DidWeb5DataV1",
  "value": Extended {
    "document": "0xa3687365727669636573a16b617470726f746f5f706473a264747970657819417470726f746f506572736f6e616c4461746153657276657268656e64706f696e747468747470733a2f2f6578616d706c652e746573746b616c736f4b6e6f776e4173816f61743a2f2f616c6963652e7465737473766572696669636174696f6e4d6574686f6473a167617470726f746f736469643a6b65793a7a5369676e696e674b6579",
    "localId": undefined,
  },
}
`;

exports[`did-web5-ts it should re-create the spec example: data1 1`] = `
Extended {
  "type": "DidWeb5DataV1",
  "value": Extended {
    "document": "0xa3687365727669636573a16b617470726f746f5f706473a264747970657819417470726f746f506572736f6e616c4461746153657276657268656e64706f696e747468747470733a2f2f6578616d706c652e746573746b616c736f4b6e6f776e4173816d61743a2f2f626f622e7465737473766572696669636174696f6e4d6574686f6473a167617470726f746f736469643a6b65793a7a5369676e696e674b6579",
    "localId": undefined,
  },
}
`;

exports[`did-web5-ts it should re-create the spec example: deactivation-tx 1`] = `
{
  "cellDeps": [
    {
      "depType": "code",
      "outPoint": {
        "index": "0x0",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
    {
      "depType": "code",
      "outPoint": {
        "index": "0x1",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
  ],
  "headerDeps": [],
  "inputs": [
    {
      "previousOutput": {
        "index": "0x0",
        "txHash": "0x781140553ba39702e39f6d8c851512549c7b930e1cb34200ab2e1b73354eccbb",
      },
      "since": "0x0",
    },
  ],
  "outputs": [
    {
      "capacity": "0x256",
      "lock": {
        "args": "0x",
        "codeHash": "0x3dc9cdb5b2dcff8d4999004406558d4456ecf0f14c757ca353684c0bd4d3b50d",
        "hashType": "type",
      },
    },
  ],
  "outputsData": [
    "0x",
  ],
  "version": "0x0",
  "witnesses": [],
}
`;

exports[`did-web5-ts it should re-create the spec example: update-tx 1`] = `
{
  "cellDeps": [
    {
      "depType": "code",
      "outPoint": {
        "index": "0x0",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
    {
      "depType": "code",
      "outPoint": {
        "index": "0x1",
        "txHash": "0x1ecbf88d692a14d7cbc0bfd1a3d5019e4b613247ae438bad52f94148c6009559",
      },
    },
  ],
  "headerDeps": [],
  "inputs": [
    {
      "previousOutput": {
        "index": "0x0",
        "txHash": "0xe54bb705cea6e500767ed4c0ba9ee92cad71e6b39c55283591147c5d10cc12b5",
      },
      "since": "0x0",
    },
  ],
  "outputs": [
    {
      "capacity": "0x257",
      "lock": {
        "args": "0x",
        "codeHash": "0x3dc9cdb5b2dcff8d4999004406558d4456ecf0f14c757ca353684c0bd4d3b50d",
        "hashType": "type",
      },
      "type": {
        "args": "0x8434cfe81aa825c275d513eee20e4235294e3420",
        "codeHash": "0xb95123c71a870e3f0f74a7ee1dab8268dbfbc1407b46733ebd1b41f854b4324a",
        "hashType": "type",
      },
    },
  ],
  "outputsData": [
    "0x00000000b20000000c000000b2000000a2000000a3687365727669636573a16b617470726f746f5f706473a264747970657819417470726f746f506572736f6e616c4461746153657276657268656e64706f696e747468747470733a2f2f6578616d706c652e746573746b616c736f4b6e6f776e4173816d61743a2f2f626f622e7465737473766572696669636174696f6e4d6574686f6473a167617470726f746f736469643a6b65793a7a5369676e696e674b6579",
  ],
  "version": "0x0",
  "witnesses": [],
}
`;

exports[`did-web5-ts it should re-create the spec example: update-tx-hash 1`] = `"0x781140553ba39702e39f6d8c851512549c7b930e1cb34200ab2e1b73354eccbb"`;
